src/Modules/OrchardCore.Commerce/Controllers/PaymentController.cs

```diff
@@ -1,75 +1,133 @@
 
     public PaymentController(
         ICardPaymentService cardPaymentService,
-        IStringLocalizer<PaymentController> stringLocalizer,
-        IAuthorizationService authorizationService,
-        IContentManager contentManager)
+        IContentItemDisplayManager contentItemDisplayManager,
+        IOrchardServices<PaymentController> services,
+        IShoppingCartHelpers shoppingCartHelpers,
+        ISiteService siteService,
+        IUpdateModelAccessor updateModelAccessor)
     {
+        _authorizationService = services.AuthorizationService.Value;
         _cardPaymentService = cardPaymentService;
-        _authorizationService = authorizationService;
-        _contentManager = contentManager;
-        T = stringLocalizer;
+        _contentItemDisplayManager = contentItemDisplayManager;
+        _logger = services.Logger.Value;
+        _contentManager = services.ContentManager.Value;
+        _shoppingCartHelpers = shoppingCartHelpers;
+        _siteService = siteService;
+        _updateModelAccessor = updateModelAccessor;
+        _userManager = services.UserManager.Value;
+        T = services.StringLocalizer.Value;
     }
 
     [Route("checkout")]
     public async Task<IActionResult> Index()
     {
-        if (User != null && !await _authorizationService.AuthorizeAsync(User, Permissions.Checkout))
+        if (!await _authorizationService.AuthorizeAsync(User, Permissions.Checkout))
         {
-            return User.Identity.IsAuthenticated ? Forbid() : LocalRedirect("~/Login?ReturnUrl=~/checkout");
+            return User.Identity?.IsAuthenticated == true ? Forbid() : LocalRedirect("~/Login?ReturnUrl=~/checkout");
         }
 
-        return View();
+        if (await _shoppingCartHelpers.CalculateSingleCurrencyTotalAsync() is not { } total) return View("CartEmpty");
+
+        var order = await _contentManager.NewAsync(CommerceContentTypes.Order);
+        var editor = await _contentItemDisplayManager.BuildEditorAsync(order, _updateModelAccessor.ModelUpdater, isNew: true);
+        var email = User.Identity?.IsAuthenticated == true
+            ? await _userManager.GetEmailAsync(await _userManager.GetUserAsync(User))
+            : null;
+
+        return View(new CheckoutViewModel
+        {
+            NewOrderEditor = editor,
+            SingleCurrencyTotal = total,
+            StripePublishableKey = (await _siteService.GetSiteSettingsAsync()).As<StripeApiSettings>().PublishableKey,
+            UserEmail = email ?? string.Empty,
+        });
     }
 
     [Route("success/{orderId}")]
     public async Task<IActionResult> Success(string orderId)
     {
-        var order = await _contentManager.GetAsync(orderId);
+        if (await _contentManager.GetAsync(orderId) is not { } order) return NotFound();
 
-        if (order == null)
-        {
-            return NotFound();
-        }
+        order.DisplayText = T["Success"].Value; // This is only for display, intentionally not saved.
```

Comment:
What is the reason?

Reply:
Don't we usually put comments above the code? I have never seen it besides it.

Reply:
The logic is that comment above applies to the whole block, comment besides applies to just that line and it can fit. This makes comment above more common, but both are valid.

Reply:
Fine then, I just didn't know about that.